src/legacy-routes.js

```diff
@@ -14,36 +19,108 @@ const invalidTx = 'Invalid transaction id!'
+  ctbId: unwrapHashPrefix(tx.hash),
+  ctbTimeIssued: moment(tx.time).unix(),
+  ctbInputs: [],
+  ctbOutputs: [],
+  ctbInputSum: { getCoin: Big(0) },
+  ctbOutputSum: { getCoin: Big(0) },
+})
+
+/**
+ * Initializes tx input/output entry
+ * @param {Movement} movement Tx input or output
+ */
+const initializeMovementEntry = (movement: Movement)
+: MovementEntry => [movement.address, { getCoin: movement.value }]
+
+/**
+ * Pushes tx movement to corresponding tx entry and sums its value
+ * according to whether it's tx input or output
+ * @param {Map<number, TxEntry>} txMap Map which holds txEntries at id keys used for fast lookups
+ * @param {Movement} movement Tx input or output
+ * @param {boolean} isInput Indicates whether to add movement to tx inputs or outputs
+ */
+const pushMovementToTxMap = (
+  txMap: Map<number, TxEntry>, movement: Movement, isInput: boolean,
+) : void => {
+  const txEntry = txMap.get(movement.txid)
+  if (!txEntry) { return }
+
+  if (isInput) {
+    txEntry.ctbInputs.push(initializeMovementEntry(movement))
+    txEntry.ctbInputSum.getCoin = txEntry.ctbInputSum.getCoin.plus(movement.value)
+  } else {
+    txEntry.ctbOutputs.push(initializeMovementEntry(movement))
+    txEntry.ctbOutputSum.getCoin = txEntry.ctbOutputSum.getCoin.plus(movement.value)
+  }
 }
 
-const txToAddressInfo = (row) => ({
-  ctbId: row.hash,
-  ctbTimeIssued: moment(row.time).unix(),
-  ctbInputs: row.inputs_address.map(
-    (addr, i) => [addr, { getCoin: row.inputs_amount[i] }]),
-  ctbOutputs: row.outputs_address.map(
-    (addr, i) => [addr, { getCoin: row.outputs_amount[i] }]),
-  ctbInputSum: {
-    getCoin: `${arraySum(row.inputs_amount)}`,
-  },
-  ctbOutputSum: {
-    getCoin: `${arraySum(row.outputs_amount)}`,
-  },
-})
+/**
+ * Assigns tx inputs and outputs to corresponding transactions to build caTxList
+ * @param {Array<Tx>} transactions Array of transactions fetched from database
+ * @param {Array<Movement>} txInputs Array of tx inputs
+ * @param {Array<Movement>} txOutputs Array of tx outputs
+ */
+const buildTxList = (
+  transactions: Array<Tx>, txInputs: Array<Movement>, txOutputs: Array<Movement>,
+) => {
+  const txMap = new Map(transactions.map(tx => [tx.id, initializeTxEntry(tx)]))
+  txInputs.forEach(txInput => pushMovementToTxMap(txMap, txInput, true))
+  txOutputs.forEach(txOutput => pushMovementToTxMap(txMap, txOutput, false))
+
+  const txList: Array<TxEntry> = [...txMap.values()]
+    .sort((a, b) => b.ctbTimeIssued - a.ctbTimeIssued)
+  return txList
+}
+
+const sumTxs = (txs, addressSet) => arraySum(txs
+  .filter(tx => addressSet.has(tx.address))
+  .map(tx => tx.value))
+
+/**
+ * An abstraction of building a result for address summary
+ * @param {*} db Database
+ * @param {*} Server Server Config Object
+ */
+const getAddressSummaryForAddresses = async (
+  dbApi: any, addresses: Array<string>,
+) => {
+  const { rows: inTxs } = await dbApi.getInwardTransactions(addresses)
```

Comment:
I think you can save this cumbersome mapping by specifying `"rowMode": "array"` in the sql query, see e.g. https://github.com/vacuumlabs/adalite-backend-service/blob/d77486bdd3aedeb3bdcbec9dcf45f224b805add2/src/db-api.js#L23

Reply:
The rowMode only transforms results from 
![image](https://user-images.githubusercontent.com/22454337/85677159-92735300-b6c7-11ea-8598-309be6366695.png)
to 
![image](https://user-images.githubusercontent.com/22454337/85677205-9e5f1500-b6c7-11ea-99d9-0418890fbee4.png)
So we have to destructure the result rows either by this unwrapping or on a separate line, the former seemed a better alternative to me.

Reply:
I see - then I would create an "extractRows" helper in the db-api and wrap the calls in it to remove that "burden" from the users of the function